xarray/core/alignment.py

```diff
@@ -322,32 +319,59 @@ def broadcast(*args):
     from .dataarray import DataArray
     from .dataset import Dataset
 
-    all_indexes = _get_all_indexes(args)
-    for k, v in all_indexes.items():
-        if not all(v[0].equals(vi) for vi in v[1:]):
-            raise ValueError('cannot broadcast arrays: the %s index is not '
-                             'aligned (use xarray.align first)' % k)
+    exclude = kwargs.pop('exclude', None)
+    if exclude is None:
+        exclude = set()
+    if kwargs:
+        raise TypeError('broadcast() got unexpected keyword arguments: %s'
+                        % list(kwargs))
+
+    args = align(*args, join='outer', copy=False, exclude=exclude)
 
     common_coords = OrderedDict()
     dims_map = OrderedDict()
     for arg in args:
         for dim in arg.dims:
-            if dim not in common_coords:
+            if dim not in common_coords and dim not in exclude:
                 common_coords[dim] = arg.coords[dim].variable
                 dims_map[dim] = common_coords[dim].size
 
     def _broadcast_array(array):
-        data = array.variable.expand_dims(dims_map)
+        # Add excluded dims to a copy of dims_map
+        array_dims_map = dims_map.copy()
+        for dim in exclude:
+            try:
+                array_dims_map[dim] = array.shape[array.dims.index(dim)]
+            except ValueError:
+                # dim not in array.dims
+                pass                
+
+        dims = tuple(array_dims_map)
+
+        data = array.variable.expand_dims(array_dims_map)
         coords = OrderedDict(array.coords)
         coords.update(common_coords)
-        dims = tuple(common_coords)
-        return DataArray(data, coords, dims, name=array.name,
-                         attrs=array.attrs, encoding=array.encoding)
+        res = DataArray(data, coords, dims, name=array.name,
+                        attrs=array.attrs, encoding=array.encoding)
+        return res
 
     def _broadcast_dataset(ds):
         data_vars = OrderedDict()
         for k in ds.data_vars:
-            data_vars[k] = ds.variables[k].expand_dims(dims_map)
+            var = ds.variables[k]
+            # Add excluded dims to a copy of dims_map    
+            var_dims_map = dims_map.copy()
+            for dim in exclude:
+                try:
+                    var_dims_map[dim] = var.shape[var.dims.index(dim)]
+                except ValueError:
+                    # dim not in var.dims
+                    try:
+                        del var_dims_map[dim]
```

Comment:
Looking at how `dims_map` is initialized, we never add add variables in `exclude`  in. So now I'm not sure how `del` could fail.